sql_metadata/parser.py

```diff
@@ -562,6 +569,45 @@ def with_names(self) -> List[str]:
         self._with_names = with_names
         return self._with_names
 
+    @property
+    def with_queries(self) -> Dict[str, str]:
+        """
+        Returns "WITH" subqueries with names
+
+        E.g. WITH tableFromWith AS (SELECT * FROM table3)
+             SELECT "xxxxx" FROM database1.tableFromWith alias
+             LEFT JOIN database2.table2 ON ("tt"."ttt"."fff" = "xx"."xxx")
+        will return {"tableFromWith": "SELECT * FROM table3"}
+        """
+        if self._with_queries is not None:
+            return self._with_queries
```

Comment:
I'm wondering if we can introduce a decorator for functions that follow this caching / memoization pattern. Instead of adding this `if` statement we can just annotate the function. But this one is for another PR 🙂

Reply:
Yes we can and I agree we should -> can be a part of the refactor with the tokens to simplify again the logic, nice idea :)